File: app/shared/src/commonMain/kotlin/org/crazydan/studio/app/hanzi/shared/StrokeFormat.kt
package org.crazydan.studio.app.hanzi.shared

/**
 * 笔画轨迹格式常量（与 web src/components/Constants.js、server/services/Trajectory.js 一致）:
 * 轨迹属性为单字符（v 版本 / b 笔刷面积比 / r 光栅实测盒 / p 坐标点）;
 * x/y 以背景汉字墨迹盒为坐标系分别归一化 ×1000（x 按盒宽、y 按盒高），
 * 压力 ×100、时间戳 ×10、笔刷面积比 ×100000 存整数
 */
object StrokeFormat {
    /** 轨迹格式版本（v2: 记录绘制时背景字光栅实测盒 r，见 server/services/Trajectory.js） */
    const val TRAJECTORY_VERSION = 2

    /** 内部坐标系尺寸（所有画布内部恒为 500×500，显示尺寸由宿主缩放） */
    const val INTERNAL_SIZE = 500

    /** 坐标归一化系数（盒相对坐标 ×1000 存整数） */
    const val COORD_SCALE = 1000f

    /** 压力归一化系数（0-1 ×100 存整数） */
    const val PRESSURE_SCALE = 100f

    /** 时间戳归一化系数（毫秒 ×10 存整数） */
    const val TIMESTAMP_SCALE = 10f

    /** 笔刷归一化系数（笔刷面积/背景字面积 比值 ×100000 存整数） */
    const val BRUSH_SCALE = 100000f

    /** 笔画间停顿（毫秒，与 web AnimationEngine strokeGap 一致） */
    const val STROKE_GAP_MS = 300f

    /** 缺省基准笔宽（内部坐标系像素，无笔刷数据时兜底） */
    const val BASE_WIDTH = 4f
}

File: shared/stroke-format.js
// ============ 笔画轨迹格式: 常量与编解码（web/server/build 单一来源） ============
// 轨迹格式由本模块权威定义，其余 JS 侧（浏览器与 Node）统一引用，禁止另行手抄数值:
//   - src/components/Constants.js（web 端重导出格式常量）
//   - server/services/Trajectory.js（服务端重导出常量并实现压缩/解压）
//   - Kotlin 端（App）无法引用 JS，其 StrokeFormat.kt/Pinyin.kt 为跨语言镜像，改动须同步
// 轨迹属性采用单字符: v 版本 / b 笔刷面积比 / r 光栅实测盒 / p 坐标点;
// x/y 以背景汉字墨迹盒为坐标系分别归一化 ×1000（x 按盒宽、y 按盒高），
// 压力 ×100、时间戳 ×10、笔刷面积比 ×100000 存整数
export const TRAJECTORY_VERSION = 2
// 坐标归一化系数（盒相对坐标 ×1000 存整数）
export const COORD_SCALE = 1000
// 压力归一化系数（0-1 ×100 存整数）
export const PRESSURE_SCALE = 100
// 时间戳归一化系数（毫秒 ×10 存整数）
export const TIMESTAMP_SCALE = 10
// 笔刷归一化: 存储值 = (笔宽² / 背景字墨迹盒面积) × BRUSH_SCALE
// 还原笔宽 = sqrt(存储值 / BRUSH_SCALE × 当前盒面积)
// 100000 精度: 典型盒下笔宽量化误差约 0.03px，远低于坐标 0.5px 粒度
export const BRUSH_SCALE = 100000
// 坐标归一化取值边界（盒相对坐标; 允许超出盒外 2 倍宽/高范围）
export const COORD_MIN = -2000
export const COORD_MAX = 3000
// 笔刷面积比取值边界（归一化比最大 1，上限 100 为兼容冗余约束，与 StrokeSchema 一致）
export const BRUSH_MIN = 0
export const BRUSH_MAX = 100 * BRUSH_SCALE

// 增量编码: 首点绝对，后续点存储与上一点的差值（时间戳单调、坐标差值小，利于压缩）
export function deltaEncode(points) {
  const out = []
  let prev = null
  for (const p of points) {
    if (!prev) {
      out.push([p[0], p[1], p[2], p[3]])
    } else {
      out.push([p[0] - prev[0], p[1] - prev[1], p[2] - prev[2], p[3] - prev[3]])
    }
    prev = p
  }
  return out
}

// 增量解码: 还原绝对坐标点
export function deltaDecode(points) {
  const out = []
  let prev = null
  for (const p of points) {
    const point = prev
      ? [p[0] + prev[0], p[1] + prev[1], p[2] + prev[2], p[3] + prev[3]]
      : [p[0], p[1], p[2], p[3]]
    out.push(point)
    prev = point
  }
  return out
}

// 点阵扁平化: [[x,y,pr,t],...] → [x,y,pr,t, ...]（每 4 个一组）
export function flattenPoints(points) {
  return points.flat()
}

// 点阵还原: [x,y,pr,t, ...] → [[x,y,pr,t],...]（每 4 个一组）
export function unflattenPoints(flat) {
  const out = []
  for (let i = 0; i + 3 < flat.length; i += 4) {
    out.push([flat[i], flat[i + 1], flat[i + 2], flat[i + 3]])
  }
  return out
}
